Fix Bug replenish + Auto Filter

ibuiv opened the pull request:
# Fix Replenish Calculation Bug + Auto Filter

## Environment
* Version: 13
* URL(s): product/stock/replenish.php

## Configuration 
MultiCompany needed with Stock Sharing ON, 2 entities and 1 warehouse per entity

**Entity A**
* Options: don't share stocks with other entities
* Warehouse : only 1 (WAR_A)
* Stock : 1 product (PROD) Qty : 4

**Entity B :**
* Options: don't share stocks with other entities
* Warehouse : only 1 (WAR_B)
* Stock : 1 product --the same-- (PROD) Qty : 17

**Product :**
* Desired Stock : 20
* Alert : 10

**WAR_A :**
* For the product
* Desired Stock : 16
* Alert : 6
 
## Fixed behavior :
* Entity A -> Replenish
* If WAR_A selected should show the product and fill with 12 the qty to order. Physical Stock (All) and Physical Stock (this 
warehouse) should be 4.

* If no Warehouse selected show the product and fill with 12 (because only one warehouse seen by this entity) the qty to order. Physical Stock (All) should be 4.

## Actual behavior :
* Entity A -> Replenish

* The product isn't showed.

* If I intentionally incrementes the Desired Stock value. The product will be shown with the good filter (Warehouse if desired stock has been incremented on the combination ProductDesired <-> Warehouse). The qty for the order are correct and the physical stock shown is correct.

## Conclusion : 
* 1 bug and 1 improvment

**BUG** : The value of the stock taken to whether or not show the product line is wrong. It is considering the stock in warehouses not shared with the entity.

**IMPROVMENT** : If only one Warehouse, take automatically the desired and alert values for this warehouse and not the general ones.

# Instructions
*This is a template to help you make good pull requests. You may use [Github Markdown](https://help.github.com/articles/getting-started-with-writing-and-formatting-on-github/) syntax to format your issue report.*
*Please:*
- *only keep the "Fix", "Close" or "New" section*
- *follow the project [contributing guidelines](/.github/CONTRIBUTING.md)*
- *replace the bracket enclosed textswith meaningful informations*


# Fix #[*issue_number Short description*]
[*Long description*]


# Close #[*issue_number Short description*]
[*Long description*]


# New [*Short description*]
[*Long description*]

ibuiv committed:
Fix Bug replenish + Auto Filter

# Fix Replenish Calculation Bug + Auto Filter

## Environment
* Version: 13
* URL(s): product/stock/replenish.php

## Configuration 
MultiCompany needed with Stock Sharing ON, 2 entities and 1 warehouse per entity

**Entity A**
* Options: don't share stocks with other entities
* Warehouse : only 1 (WAR_A)
* Stock : 1 product (PROD) Qty : 4

**Entity B :**
* Options: don't share stocks with other entities
* Warehouse : only 1 (WAR_B)
* Stock : 1 product --the same-- (PROD) Qty : 17

**Product :**
* Desired Stock : 20
* Alert : 10

**WAR_A :**
* For the product
* Desired Stock : 16
* Alert : 6
 
## Fixed behavior :
* Entity A -> Replenish
* If WAR_A selected should show the product and fill with 12 the qty to order. Physical Stock (All) and Physical Stock (this 
warehouse) should be 4.

* If no Warehouse selected show the product and fill with 12 (because only one warehouse seen by this entity) the qty to order. Physical Stock (All) should be 4.

## Actual behavior :
* Entity A -> Replenish

* The product isn't showed.

* If I intentionally incrementes the Desired Stock value. The product will be shown with the good filter (Warehouse if desired stock has been incremented on the combination ProductDesired <-> Warehouse). The qty for the order are correct and the physical stock shown is correct.

## Conclusion : 
* 1 bug and 1 improvment

**BUG** : The value of the stock taken to whether or not show the product line is wrong. It is considering the stock in warehouses not shared with the entity.

**IMPROVMENT** : If only one Warehouse, take automatically the desired and alert values for this warehouse and not the general ones.

stickler-ci committed:
Fixing style errors.

eldy committed:
Update replenish.php

eldy committed:
Update replenish.php

eldy reviewed (commented):
Restore first suggestion to comment it

eldy commented on htdocs/product/stock/replenish.php:
```diff
@@ -340,8 +340,8 @@
 $sql .= ' FROM '.MAIN_DB_PREFIX.'product as p';
-$sql .= ' LEFT JOIN '.MAIN_DB_PREFIX.'entrepot AS ent ON ent.entity IN('.getEntity('stock').')';
-$sql .= ' LEFT JOIN '.MAIN_DB_PREFIX.'product_stock as s ON p.rowid = s.fk_product AND s.fk_entrepot = ent.rowid';
+$sql .= ' LEFT JOIN '.MAIN_DB_PREFIX.'product_stock as s ON p.rowid = s.fk_product';
```
The new sql does not look good because first join seem to be a cartesien product with no link and filter is done after.   I suggest tho change in a better way to understand the request, like this :

$listofqualifiedwarehousesid = get list of warehouse id with entity in getEntity('stock')

Then 

$sql .= ' FROM '.MAIN_DB_PREFIX.'product as p';
$sql .= ' LEFT JOIN '.MAIN_DB_PREFIX.'product_stock as s ON p.rowid = s.fk_product AND s.fk_entrepot IN ('.$db->sanitize($listofqualifiedwarehousesid).')';
//$sql .= ' LEFT JOIN '.MAIN_DB_PREFIX.'entrepot AS ent ON s.fk_entrepot = ent.rowid AND ent.entity IN('.getEntity('stock').')';

ibuiv replied:
All your remarks have been applied. 
Thank you very much for your advices

ibuiv committed:
lint (code) Apply remarks and upgrade codee

- Sets List of all Visible Warehouses at the beginning with getEntity method (removed the call to global $mc)
- Better check for warehouse auto select (`$count == 1 && (empty($fk_entrepot) || $fk_entrepot <= 0) && ...`)
- Better SQL applying eldy's remarks
